src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs

```diff
@@ -766,6 +782,14 @@ private class Dropbox
             private readonly ConcurrentDictionary<int, TaskCompletionSource<IByteBuffer>> _messages = new ConcurrentDictionary<int, TaskCompletionSource<IByteBuffer>>();
 
             // public methods
+            public void AddException(Exception exception)
+            {
+                foreach (var tsc in _messages.Values)
+                {
+                    tsc.TrySetException(exception);
```

Comment:
Maybe add a comment:

    tsc.TrySetException(exception); // has no effect on already completed tasks

Reply:
You asked why we have to set the exception on all pending tasks...

If this method is called the connection has failed (you might even rename this method to `ConnectionFailed`), and all pending tasks are now doomed and that should be signaled by setting all of them to failed with the corresponding exception.

Since we don't currently do pipelining in practice there should only be one pending task in the dropbox at a time, but you might as well code it correctly.

I still think the better solution in the long run is to remove the Dropbox. Its operation is too subtle and its functionality is not needed since we don't intend to ever support pipelining (we once did but we no longer do).

Reply:
Added the comment. This solution works for multiple threads accessing the `BinaryConnection`